fix: disable parallel build when extractCSS is enabled

stevefan1999-personal opened the pull request:
<!--- Provide a general summary of your changes in the title above -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (perhaps, it is a breaking change for the people who are broke with this before)


## Description
Fixes https://github.com/nuxt/nuxt.js/issues/4965#issuecomment-462375082, details can be seen in the comments

## Checklist:
<!--- Put an `x` in all the boxes that apply. -->
<!--- If your change requires a documentation PR, please link it appropriately -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

<!-- *Sigh* I shouldn't work using github edit and really should clone it locally -->

stevefan1999-personal committed:
Update options.js

clarkdo commented:
@stevefan1999-personal Thank you for the pr.

About the `thread-loader` context issue, I actually met the similar one which is solved by adding properties to `thread-loader` context, how do you think fix it in same way ?

https://github.com/webpack-contrib/thread-loader/pull/25

manniL commented:
Please fix the lint issues ☺️

stevefan1999-personal committed:
fix: no-trailing-spaces

stevefan1999-personal commented:
@clarkdo Nope, you still cannot access the rest of the webpack context:
https://github.com/faceyspacey/extract-css-chunks-webpack-plugin/blob/43e1672cd231301581df929a4159c7b716e64065/src/loader.js#L43-L46

So, unless `thread-loader` did an architectural change which allows master-worker communication, this will be philosophically impossible to fix.

clarkdo commented:
@stevefan1999-personal Could you fix the broken code shown in failed test ?

manniL commented on packages/config/src/options.js:
```diff
@@ -281,6 +281,11 @@ export function getNuxtConfig(_options) {
     consola.warn('vendor has been deprecated due to webpack4 optimization')
   }
 
+  if (options.build.parallel && typeof options.build.extractCSS !== 'undefined') {
```
```suggestion
  if (options.build && options.build.parallel && typeof options.build.extractCSS !== 'undefined') {
```

Maybe? ☺️

stevefan1999-personal replied:
Although it is rare to not have `build` option, but still a potential read hazard, kudos.

TheAlexLichter committed:
Update packages/config/src/options.js

Co-Authored-By: stevefan1999-personal <29133953+stevefan1999-personal@users.noreply.github.com>

stevefan1999-personal committed:
Merge branch 'dev' into patch-parallel-extract-css

clarkdo commented:
Actually error is here https://github.com/nuxt/nuxt.js/blob/master/packages/config/src/options.js#L285 since `extractCSS` is deleted

stevefan1999-personal committed:
fix: reading undeinfed options.build.extractCSS

stevefan1999-personal commented:
@clarkdo what are the reasons those `typeof ... === 'undefined'` are needed instead of boolean coerce?

clarkdo commented:
@stevefan1999-personal There is no much difference in this case, you can use both. 

Performance is only slightly different https://jsperf.com/type-of-undefined-vs-undefined/30

clarkdo commented:
Now the issue is that this pr changed a little bit allchunks warn message in test/fixtures/with-config since it's using parallel build and extactCSS, maybe we should move the allchunks warn message test to extractcss fixtures.

clarkdo commented on packages/config/src/options.js:
```diff
@@ -281,11 +281,17 @@ export function getNuxtConfig(_options) {
   }
 
+  if (options.build && options.build.parallel && typeof options.build.extractCSS !== 'undefined') {
+    delete options.build.extractCSS
```
This may be simple and no need to other checks
```suggestion
options.build.extractCSS = false
```

clarkdo commented on packages/config/src/options.js:
```diff
@@ -281,14 +281,14 @@ export function getNuxtConfig(_options) {
   // build.extractCSS.allChunks has no effect
-  if (typeof options.build.extractCSS !== 'undefined' &&
-      typeof options.build.extractCSS.allChunks !== 'undefined') {
+  if (options.build.extractCSS && options.build.extractCSS.allChunks) {
```
what is options.build.extractCSS.allChunks = false ?

clarkdo commented on packages/config/src/options.js:
```diff
@@ -281,8 +281,14 @@ export function getNuxtConfig(_options) {
+
   // build.extractCSS.allChunks has no effect
-  if (typeof options.build.extractCSS.allChunks !== 'undefined') {
+  if (typeof options.build.extractCSS === 'object' && options.build.extractCSS.allChunks) {
```
I think you can just revert this part since `extractCSS` is not undefined by `false` after your previous change

clarkdo commented:
Sorry for the so many comments 🤖

stevefan1999-personal committed:
chore: squashing commits

chore(examples): fix links to vue-meta repo (#5018)


chore(deps): update dependency vue-jest to ^4.0.0-beta.1 (#5016)


fix: set extractCSS to false instead of deletion
Merge branch 'dev' into patch-parallel-extract-css
chore: squashing commits

update: re-add the missing new parallel log

add: missing comments

fix: type assertion for extractCSS

revert: unnecessary checks

revert: test mock count and extract CSS

stevefan1999-personal committed:
Merge branch 'dev' into patch-parallel-extract-css

stevefan1999-personal commented:
Okay. So I squashed those commits in hope to restore the CI test and it fails me again and again. I don't know why are there 503s and 500s. I'm completely speechless now.

clarkdo commented:
The test error is because extractcss is turned off in parallel mode.

But in with-config, it’s built with parallel mode and extractcss, so the expected warning message disappeared with your pr change, I’ve suggested that move this part test to extractcss fixture😺

clarkdo commented:
No worries, I’ll do the test chages for you, the current change LGTM

williamchong007 commented:
Just my 2 cents
Shouldn't `extractCSS` be kept and `parallel` be disabled instead?
css affects the end product, and parallel just affects build time performance
Should we prefer to keep what user expect to get as a result, instead of a performance setting?

clarkdo committed:
unit test

clarkdo commented:
@williamchong007 Good idea, I agree with you

clarkdo committed:
fix: disable parallel
